lib/luaxtable.c

```diff
@@ -0,0 +1,232 @@
+
+typedef enum luaxtable_type_e {
+	LUAXTABLE_TMATCH,
+	LUAXTABLE_TTARGET,
+} luaxtable_type_t;
+
+typedef struct luaxtable_s {
+	lunatik_object_t *runtime;
+	union {
+		struct xt_match match;
+		struct xt_target target;
+	};
+	luaxtable_type_t type;
+	bool registered;
+} luaxtable_t;
+
+static bool luaxtable_match(const struct sk_buff *skb, struct xt_action_param *par)
+{
+	return true;
+}
+
+static int luaxtable_match_check(const struct xt_mtchk_param *par)
+{
+	return 0;
+}
+
+static void luaxtable_match_destroy(const struct xt_mtdtor_param *par)
+{
+}
+
+static unsigned int luaxtable_target(struct sk_buff *skb, const struct xt_action_param *par)
+{
+	return NF_ACCEPT;
+}
+
+static int luaxtable_target_check(const struct xt_tgchk_param *par)
+{
+	return 0;
+}
+
+static void luaxtable_target_destroy(const struct xt_tgdtor_param *par)
+{
+}
+
+static void luaxtable_release(void *private);
+
+static const luaL_Reg luaxtable_mt[] = {
+	{"__gc", lunatik_deleteobject},
+	{NULL, NULL}
+};
+
+static const lunatik_class_t luaxtable_class = {
+	.name = "xtable",
+	.methods = luaxtable_mt,
+	.release = luaxtable_release,
+	.sleep = false,
+};
+
+#define luaxtable_setinteger(L, idx, hook, field) 		\
+do {								\
+	lunatik_checkfield(L, idx, #field, LUA_TNUMBER);	\
+	hook->field = lua_tointeger(L, -1);			\
+	lua_pop(L, 1);						\
+} while (0)
+
+#define luaxtable_setstring(L, idx, hook, field, maxlen)        \
+do {								\
+	const char *str;					\
+	size_t len;						\
+								\
+	lunatik_checkfield(L, idx, #field, LUA_TSTRING);	\
+	str = lua_tolstring(L, -1, &len);			\
+	if (len >= maxlen)					\
+		luaL_error(L, "'%s' is too long", #field);	\
+	strncpy((char *)hook->field, str, maxlen);		\
+	lua_pop(L, 1);						\
+} while (0)
+
+static inline lunatik_object_t *luaxtable_new(lua_State *L, int hook)
+{
+	lunatik_object_t *object = lunatik_newobject(L, &luaxtable_class , sizeof(luaxtable_t));
+	luaxtable_t *xtable = (luaxtable_t *)object->private;
+
+	xtable->type = hook;
+	xtable->registered = false;
+	return object;
+}
+
+#define LUAXTABLE_NEWHOOK(hook, HOOK)					\
+static int luaxtable_new##hook(lua_State *L) 				\
+{									\
+	luaL_checktype(L, 1, LUA_TTABLE);				\
+									\
+	lunatik_object_t *object = luaxtable_new(L, HOOK); 		\
+	luaxtable_t *xtable = (luaxtable_t *)object->private;		\
+									\
+	struct xt_##hook *hook = &xtable->hook;				\
+	hook->me = THIS_MODULE;						\
+									\
+	luaxtable_setstring(L, 1, hook, name, XT_EXTENSION_MAXNAMELEN);	\
+	luaxtable_setinteger(L, 1, hook, revision);			\
+	luaxtable_setinteger(L, 1, hook, family);			\
+	luaxtable_setinteger(L, 1, hook, proto);			\
+									\
+	hook->hook = luaxtable_##hook;					\
+	hook->checkentry = luaxtable_##hook##_check;			\
+	hook->destroy = luaxtable_##hook##_destroy;			\
+									\
+	if (xt_register_##hook(hook) != 0)				\
+		luaL_error(L, "unable to register " #hook);		\
+									\
+	xtable->registered = true;					\
+	xtable->runtime = lunatik_toruntime(L);				\
```

Comment:
```suggestion
	xtable->runtime = lunatik_toruntime(L);				\
```
we can actual use `xtable->runtime` for this purpose.. now I see.. =)..